sql: explain bundle doesn't work with dbeaver

```SQL
explain bundle select * from drop_index_test;
```
![image](https://user-images.githubusercontent.com/22278911/76483621-dfb80700-63ed-11ea-9316-c76f2abd27f0.png)

Here is the error:
```
SQL Error [0A000]: ERROR: unimplemented: unsupported statement: *tree.ExplainBundle
  Hint: You have attempted to use a feature that is not yet implemented.
See: https://github.com/cockroachdb/cockroach/issues/34848
```
```
org.jkiss.dbeaver.model.sql.DBSQLException: SQL Error [0A000]: ERROR: unimplemented: unsupported statement: *tree.ExplainBundle
  Hint: You have attempted to use a feature that is not yet implemented.
See: https://github.com/cockroachdb/cockroach/issues/34848
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:134)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeStatement(SQLQueryJob.java:488)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.lambda$0(SQLQueryJob.java:425)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:170)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeSingleQuery(SQLQueryJob.java:417)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.extractData(SQLQueryJob.java:775)
	at org.jkiss.dbeaver.ui.editors.sql.SQLEditor$QueryResultsContainer.readData(SQLEditor.java:2914)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.lambda$0(ResultSetJobDataRead.java:111)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:170)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.run(ResultSetJobDataRead.java:109)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer$17.run(ResultSetViewer.java:3423)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:103)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: org.postgresql.util.PSQLException: ERROR: unimplemented: unsupported statement: *tree.ExplainBundle
  Hint: You have attempted to use a feature that is not yet implemented.
See: https://github.com/cockroachdb/cockroach/issues/34848
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2440)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2183)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:308)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:307)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:293)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:270)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:266)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.execute(JDBCStatementImpl.java:338)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:131)
	... 12 more
```
Explain analyze does work for example:
![image](https://user-images.githubusercontent.com/22278911/76483632-ea729c00-63ed-11ea-8261-bc9ad360a323.png)

Version: V20.1.0-BETA.2-764-G73F4AAC551
Dbeaver: Version 7.0.0.202003021717

To repro:
Use dbeaver with cockroach demo.
Then use the SQL editor to execute explain bundle.

--- a/pkg/sql/conn_executor_exec.go
+++ b/pkg/sql/conn_executor_exec.go
@@ -215,6 +215,11 @@ func (ex *connExecutor) execStmtInOpenState(
 		stmt.AST = explainBundle.Statement
 		// TODO(radu): should we trim the "EXPLAIN BUNDLE" part from stmt.SQL?
 
+		// Clear any ExpectedTypes we set if we prepared this statement (they
+		// reflect the column types of the EXPLAIN itself and not those of the inner
+		// statement).
+		stmt.ExpectedTypes = nil
+
 		// EXPLAIN BUNDLE does not return the rows for the given query; instead it
 		// returns some text which includes a URL.
 		// TODO(radu): maybe capture some of the rows and include them in the

--- a/pkg/sql/explain_bundle_test.go
+++ b/pkg/sql/explain_bundle_test.go
@@ -38,17 +38,50 @@ func TestExplainBundle(t *testing.T) {
 	defer srv.Stopper().Stop(ctx)
 	r := sqlutils.MakeSQLRunner(godb)
 
-	r.Exec(t, "CREATE TABLE abc (a INT PRIMARY KEY, b INT, c INT UNIQUE)")
-	rows := r.QueryStr(t, "EXPLAIN BUNDLE SELECT * FROM abc WHERE c=1")
-	checkBundle(t, fmt.Sprint(rows), "statement.txt opt.txt opt-v.txt opt-vv.txt plan.txt trace.json")
+	// Expected list of files for a successful statement.
+	expSuccess := "statement.txt opt.txt opt-v.txt opt-vv.txt plan.txt trace.json"
+	// Expected list of files for a statement that fails during planning.
+	expPlanFail := "statement.txt trace.json"
 
-	// Even on query errors, we should still get a bundle.
-	_, err := godb.QueryContext(ctx, "EXPLAIN BUNDLE SELECT * FROM badtable")
-	if !testutils.IsError(err, "relation.*does not exist") {
-		t.Fatalf("unexpected error %v\n", err)
-	}
-	// The bundle url is inside the error detail.
-	checkBundle(t, fmt.Sprintf("%+v", err.(*pq.Error).Detail), "statement.txt trace.json")
+	t.Run("basic", func(t *testing.T) {
+		r.Exec(t, "CREATE TABLE abc (a INT PRIMARY KEY, b INT, c INT UNIQUE)")
+		rows := r.QueryStr(t, "EXPLAIN BUNDLE SELECT * FROM abc WHERE c=1")
+		checkBundle(t, fmt.Sprint(rows), expSuccess)
+	})
+
+	// Even on query errors we should still get a bundle.
+	t.Run("error", func(t *testing.T) {
+		_, err := godb.QueryContext(ctx, "EXPLAIN BUNDLE SELECT * FROM badtable")
+		if !testutils.IsError(err, "relation.*does not exist") {
+			t.Fatalf("unexpected error %v\n", err)
+		}
+		// The bundle url is inside the error detail.
+		checkBundle(t, fmt.Sprintf("%+v", err.(*pq.Error).Detail), expPlanFail)
+	})
+
+	// Verify that we can issue the statement with prepare (which can happen
+	// depending on the client).
+	t.Run("prepare", func(t *testing.T) {
+		stmt, err := godb.Prepare("EXPLAIN BUNDLE SELECT * FROM abc WHERE c=1")
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer stmt.Close()
+		rows, err := stmt.Query()
+		if err != nil {
+			t.Fatal(err)
+		}
+		var rowsBuf bytes.Buffer
+		for rows.Next() {
+			var row string
+			if err := rows.Scan(&row); err != nil {
+				t.Fatal(err)
+			}
+			rowsBuf.WriteString(row)
+			rowsBuf.WriteByte('\n')
+		}
+		checkBundle(t, rowsBuf.String(), expSuccess)
+	})
 }
 
 // checkBundle searches text strings for a bundle URL and then verifies that the

--- a/pkg/sql/opt/optbuilder/builder.go
+++ b/pkg/sql/opt/optbuilder/builder.go
@@ -303,6 +303,10 @@ func (b *Builder) buildStmt(
 	case *tree.Export:
 		return b.buildExport(stmt, inScope)
 
+	case *tree.ExplainBundle:
+		// This statement should have been handled by the executor.
+		panic(errors.Errorf("%s can only be used as a top-level statement", stmt.StatementTag()))
+
 	default:
 		// See if this statement can be rewritten to another statement using the
 		// delegate functionality.
@@ -325,7 +329,7 @@ func (b *Builder) buildStmt(
 			b.DisableMemoReuse = true
 			return outScope
 		}
-		panic(unimplementedWithIssueDetailf(34848, stmt.StatementTag(), "unsupported statement: %T", stmt))
+		panic(errors.AssertionFailedf("unexpected statement: %T", stmt))
 	}
 }
 

--- a/pkg/sql/plan_opt.go
+++ b/pkg/sql/plan_opt.go
@@ -25,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlbase"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
 )
 
 var queryCacheEnabled = settings.RegisterBoolSetting(
@@ -43,13 +44,8 @@ func (p *planner) prepareUsingOptimizer(ctx context.Context) (planFlags, error) 
 	opc := &p.optPlanningCtx
 	opc.reset()
 
-	// These statements do not have result columns and do not support placeholders
-	// so there is no need to do anything during prepare.
-	//
-	// Some of these statements (like BeginTransaction) aren't supported by the
-	// optbuilder so they would error out. Others (like CreateIndex) have planning
-	// code that can introduce unnecessary txn retries (because of looking up
-	// descriptors and such).
+	stmt.Prepared.AnonymizedStr = anonymizeStmt(stmt.AST)
+
 	switch stmt.AST.(type) {
 	case *tree.AlterIndex, *tree.AlterTable, *tree.AlterSequence,
 		*tree.BeginTransaction,
@@ -68,6 +64,22 @@ func (p *planner) prepareUsingOptimizer(ctx context.Context) (planFlags, error) 
 		*tree.RollbackToSavepoint, *tree.RollbackTransaction,
 		*tree.Savepoint, *tree.SetTransaction, *tree.SetTracing, *tree.SetSessionAuthorizationDefault,
 		*tree.SetSessionCharacteristics:
+		// These statements do not have result columns and do not support placeholders
+		// so there is no need to do anything during prepare.
+		//
+		// Some of these statements (like BeginTransaction) aren't supported by the
+		// optbuilder so they would error out. Others (like CreateIndex) have planning
+		// code that can introduce unnecessary txn retries (because of looking up
+		// descriptors and such).
+		return opc.flags, nil
+
+	case *tree.ExplainBundle:
+		// This statement returns result columns but does not support placeholders,
+		// and we don't want to do anything during prepare.
+		if len(p.semaCtx.Placeholders.Types) != 0 {
+			return 0, errors.Errorf("%s does not support placeholders", stmt.AST.StatementTag())
+		}
+		stmt.Prepared.Columns = sqlbase.ExplainBundleColumns
 		return opc.flags, nil
 	}
 
@@ -101,8 +113,6 @@ func (p *planner) prepareUsingOptimizer(ctx context.Context) (planFlags, error) 
 		}
 		opc.flags.Set(planFlagOptCacheMiss)
 	}
-
-	stmt.Prepared.AnonymizedStr = anonymizeStmt(stmt.AST)
 
 	memo, err := opc.buildReusableMemo(ctx)
 	if err != nil {

